haskell/cabal.bzl

```diff
@@ -2237,6 +2314,7 @@ _stack_snapshot = repository_rule(
         "verbose": attr.bool(default = False),
         "custom_toolchain_libraries": attr.string_list(default = []),
         "enable_custom_toolchain_libraries": attr.bool(default = False),
+        "components_args": attr.string_dict(),
```

Comment:
Likewise, could we use `attr.label_keyed_string_dict(providers = [HaskellCabalLibraryArgs])` here instead, with a TODO to switch to `string_keyed_label_dict` some time later.

This makes it explicit what to expect here.

Just call `_invert` before passing its value to the `_stack_snapshot` rule.